### PR TITLE
Use backtrace from dmlc-core (dmlc::stack_trace())

### PR DESCRIPTION
**Pull request by larroy**
remove code duplication
demangle C++ symbols

### CODE REVIEW

**Comment by larroy** [comment]
https://github.com/dmlc/dmlc-core/pull/336

**Comment by cjolivier01** [comment]
On other PR it was decided that printing dmlc stacktrace function was desired? I remember you mentioning that withjout a macro this would happen:
 /workspace/scala-package/init-native/linux-x86_64/target/libmxnet-init-scala-linux-x86_64.so(_ZN4dmlc10StackTraceB5cxx11Ev+0x1bd

File: make/config.mk [hunk 1]
```diff
@@ -33,8 +33,8 @@ DEBUG = 0
 # whether compile with profiler
 USE_PROFILER =
 
-# whether to turn on signal handler (e.g. segfault logger)
-USE_SIGNAL_HANDLER =
+# whether to turn on segfault signal handler to log the stack trace
+USE_SEGFAULT_HANDLER = 1
```
Comment (piiswrong) [inline make/config.mk]: Don't change the name.

Comment (piiswrong) [reply make/config.mk]: we plan to add FPE too.

Comment (larroy) [reply make/config.mk]: Done

File: src/initialize.cc [hunk 2]
```diff
@@ -25,37 +25,23 @@
...
-  fprintf(stderr, "Stack trace returned %d entries:\n", nframes);
-  char **msgs = backtrace_symbols(stack, nframes);
-  if (msgs != nullptr) {
-    for (int i = 0; i < nframes; ++i) {
-      fprintf(stderr, "[bt] (%d) %s\n", i, msgs[i]);
-    }
-  }
-#endif  // DMLC_LOG_STACK_TRACE
-
+  fprintf(stderr, "%s", dmlc::StackTrace().c_str());
```
Comment (cjolivier01) [inline src/initialize.cc]: Will fprintf() do a flush on \n or should fflush() be called here?

Comment (piiswrong) [reply src/initialize.cc]: Is `dmlc::StackTrace().c_str()` safe? wouldn't the pointer be invalid because the returned string will be released?

Comment (larroy) [reply src/initialize.cc]: @piiswrong it's safe, temporaries last until the end of the current statement.

Comment (larroy) [reply src/initialize.cc]: @cjolivier01 stderr is unbuffered in *nix https://linux.die.net/man/3/stderr  it does a flush in \n

**Comment by larroy** [comment]
@cjolivier01 yes, it can be tested.

**Comment by piiswrong** [comment]
You are not updating dmlc-core in this PR. Is it already updated and contains StackTrace?

**Comment by larroy** [comment]
Yes, I checked, and CI wouldn't pass otherwise.

**Pull request merged**